src/modules/mavlink/mavlink_messages.cpp

```diff
@@ -5243,6 +5244,172 @@ class MavlinkStreamRawRpm : public MavlinkStream
+	{}
+
+	bool send(const hrt_abstime t) override
+	{
+
+		esc_status_s esc_status;
+
+		uint8_t batch_size = MAVLINK_MSG_ESC_INFO_FIELD_TEMPERATURE_LEN;
+
+		if (_esc_status_sub.update(&esc_status)) {
+
+			mavlink_esc_info_t msg = {};
+
+			msg.time_usec = esc_status.timestamp;
+			msg.counter = esc_status.counter;
+			msg.count = esc_status.esc_count;
+			msg.connection_type = esc_status.esc_connectiontype;
+			msg.info = esc_status.esc_online_flags;
+
+			uint8_t number_batches = (esc_status.esc_count + (batch_size - 1)) / batch_size;
+
+			for (int batch_number = 0; batch_number < number_batches; batch_number++) {
+
+				msg.index = batch_number * batch_size;
+
+				for (int esc_index = 0; esc_index < batch_size ; esc_index++) {
+
+					msg.failure_flags[esc_index] = esc_status.esc[esc_index].failures;
+					msg.error_count[esc_index] = esc_status.esc[esc_index].esc_errorcount;
+					msg.temperature[esc_index] = esc_status.esc[esc_index].esc_temperature;
+				}
+
+				mavlink_msg_esc_info_send_struct(_mavlink->get_channel(), &msg);
+
+			}
+
+			return true;
+		}
+
+		return false;
+	}
+};
+
+class MavlinkStreamESCStatus : public MavlinkStream
+{
+public:
+	const char *get_name() const override
+	{
+		return MavlinkStreamESCStatus::get_name_static();
+	}
+
+	static constexpr const char *get_name_static()
+	{
+		return "ESC_STATUS";
+	}
+
+	static constexpr uint16_t get_id_static()
+	{
+		return MAVLINK_MSG_ID_ESC_STATUS;
+	}
+
+	uint16_t get_id() override
+	{
+		return get_id_static();
+	}
+
+	static MavlinkStream *new_instance(Mavlink *mavlink)
+	{
+		return new MavlinkStreamESCStatus(mavlink);
+	}
+
+	unsigned get_size() override
+	{
+		return _esc_status_sub.advertised() ? MAVLINK_MSG_ID_ESC_STATUS_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES : 0;
+	}
+
+private:
+	uORB::Subscription _esc_status_sub{ORB_ID(esc_status)};
+
+	/* do not allow top copying this class */
+	MavlinkStreamESCStatus(MavlinkStreamESCStatus &) = delete;
+	MavlinkStreamESCStatus &operator = (const MavlinkStreamESCStatus &) = delete;
+
+protected:
+	explicit MavlinkStreamESCStatus(Mavlink *mavlink) : MavlinkStream(mavlink)
+	{}
+
+	bool send(const hrt_abstime t) override
+	{
+
+		esc_status_s esc_status;
+		uint8_t batch_size = MAVLINK_MSG_ESC_STATUS_FIELD_RPM_LEN;
+
+		if (_esc_status_sub.update(&esc_status)) {
+
+			mavlink_esc_status_t msg = {};
+
+			msg.time_usec = esc_status.timestamp;
+
+			uint8_t number_batches = (esc_status.esc_count + (batch_size - 1)) / batch_size;
+
+			for (int batch_number = 0; batch_number < number_batches; batch_number++) {
+
+				msg.index = batch_number * batch_size;
+
+				for (int esc_index = 0; esc_index < batch_size ; esc_index++) {
+
+					msg.rpm[esc_index] = esc_status.esc[esc_index].esc_rpm;
+					msg.voltage[esc_index] = esc_status.esc[esc_index].esc_voltage;
+					msg.current[esc_index] = esc_status.esc[esc_index].esc_current;
+				}
+
+				mavlink_msg_esc_status_send_struct(_mavlink->get_channel(), &msg);
```

Comment:
The `get_size()` method should try to be accurate as the mavlink module is checking this across all active streams to set the rate multiplier as the connection degrades.

Count you store count (esc_status.esc_count) and use it in `get_size()` to return the size of the data that's actually going to be sent?

Reply:
Can you check again @dagar ?

I also rebased.

Reply:
Looks right.